load.c

```diff
@@ -328,6 +337,15 @@ get_loaded_features_index(void)
 	    features_index_add(as_str, INT2FIX(i));
 	}
 	reset_loaded_features_snapshot();
+
+        features = rb_ary_dup(vm->loaded_features_snapshot);
```

Comment:
Ok, so after digging a bit more, this path should only be taken when `$LOADED_FEATURES` has been mutated from the ruby side, so it might be caused by Bootsnap's `loaded_features_index`.

I'll try to confirm this and to see if there's a way to avoid that.

Reply:
Hum, actually, I don't think bootsnap mutates `$LOADED_FEATURES`, so I'm not sure what cause the index to be rebuilt so much. I'll have to dig more.

Reply:
I'm at a loss here, I used the following monkey patch to try to figure out what might be mutating `$LOADED_FEATURES`:

```ruby
module LoadedObserver
  ($LOADED_FEATURES.methods - [:any?, :size, :last, :each, :each_with_object, :reverse_each, :[]]).each do |m|
    module_eval(<<~RUBY, __FILE__, __LINE__ + 1)
      def #{m}(...)
        puts "-" * 40
        p [:call, "#{m}"]
        puts caller
        super
      end
    RUBY
  end
end

$LOADED_FEATURES.singleton_class.prepend(LoadedObserver)
```

But I got absolutely no hits.

Reply:
Ok, so I went as far as modifying Ruby, and I was wrong, this callsite isn't where the arrays are created.

However looking at the heap dumps, I can confirm that these arrays are indeed some `$LOAD_PATH` snapshots, I just don't get how they are created.

I suspect it might have something to do with `reset_loaded_features_snapshot()`, but I must confess I don't quite grasp how shared arrays work exactly.